Fix childprocess killing when the parent process SIGINTs

TimvdLippe opened the pull request:
If you `ctrl + c` the Puppeteer parent process, we would sometimes
not properly handle killing of the child processes. This would then
leave child processes behind, with running Chromium instances. This
in turn could block Puppeteer from launching again and results in
cryptic errors.

Instead of using the generic `process.kill` with the process id
(which for some reason is negative the pid, which I don't get),
we can kill the child process directly by calling `proc.kill`.

Additionally, we have to remove the eventlisteners as the last
invocation of a Node emitter callback. If we remove the listener
before the full callback is run, it would prevent further execution.
E.g. if the first call is the removal of the event listener, then the
cleanup later in the callback would not happen.

You can test this locally by running the unittests (`npm run unit`)
and terminating the process prematurely (`ctrl + c`). Before this
change, the process would terminate with exit code 0. With this
change, it now properly terminates with exit code 130 (as defined
by the `BrowserRunner.ts` in the `SIGINT` callback).

Lastly, I added a more helpful error message in the event that
Puppeteer is not able to clean up the Node process. In theory,
this should now no longer happen. However, if it does, we now
explain what is going on and where users can report their bug.

After this change, whenever you run the unit tests or anything
related to Puppeteer, you should have zero browser processes
lingering. E.g. `ps aux | grep Chromium` should show no
results. This therefore prevents the need for continuous killing
of the browser processes when working on Puppeteer.

Fixes #5729
Fixes #4796
Fixes #4963
Fixes #4333
Fixes #1825

TimvdLippe committed:
Fix childprocess killing when the parent process SIGINTs

If you `ctrl + c` the Puppeteer parent process, we would sometimes
not properly handle killing of the child processes. This would then
leave child processes behind, with running Chromium instances. This
in turn could block Puppeteer from launching again and results in
cryptic errors.

Instead of using the generic `process.kill` with the process id
(which for some reason is negative the pid, which I don't get),
we can kill the child process directly by calling `proc.kill`.

TimvdLippe commented:
This is an attempt to fix the pesky "`ctrl + C` and now my Puppeteer is broken" issue. Pushing this to Travis to see if it is happy. I will need to do more testing locally verify it is actually fixed. Would be great if you could test this too!

TimvdLippe commented:
Seems like there is only 1 test failing. That seems hopeful. There are some failures on Windows though. But I suppose we can use the same `proc.kill` there.

jackfranklin commented:
The test failures look like a simple error message has changed too which is great.

Could we write a test that demonstrates this new behaviour?

TimvdLippe committed:
Always kill proc on Windows and add nicer error message

TimvdLippe committed:
Merge branch 'master' into fix-proc-kill

TimvdLippe committed:
Fix SIGINT handler

TimvdLippe committed:
Relax check for Windows

On Windows, it might not always cleanup the process, as it was
already cleaned up. We can relax the check here.

TimvdLippe committed:
Merge branch 'master' into fix-proc-kill

mathiasbynens commented on mocha-config/base.js:
```diff
@@ -1,4 +1,6 @@
 module.exports = {
   reporter: 'dot',
+  // Allow `console.logs` to show up during test execution
```
nit:

```suggestion
  // Allow `console.log`s to show up during test execution
```

mathiasbynens commented on src/launcher/BrowserRunner.ts:
```diff
@@ -131,24 +134,32 @@ export class BrowserRunner {
+        this.proc.kill('SIGKILL');
+      } catch (error) {
+        throw new Error(`${PROCESS_ERROR_EXPLANATION}
+Error cause: ${error.stack}`);
```
nit to maintain indentation:

```suggestion
        throw new Error(`${PROCESS_ERROR_EXPLANATION}\nError cause: ${error.stack}`);
```

mathiasbynens reviewed (approved):
LGTM! Thanks for looking into this 🎉

TimvdLippe committed:
Fix nits

TimvdLippe committed:
Fix formatter
